feat: add PPXA and ConsensusADMM algorithms

tttamaki opened the pull request:
I added the following two optimizers, both for minimizing the sum of prox operators.

- Parallel Proximal Algorithm (PPXA) for minimizing the sum of proximable functions
- ConsensusADMM for solving the global consensus problem with ADMM

also I added some tests with existing ADMM and GeneralizedProximalGradient.

The function of these optimizers is similar to GeneralizedProximalGradient, but more common in books and hence better to be implemented.

tttamaki committed:
add PPXA and ConsensusADMM algorithms with corresponding documentation and tests

- Parallel Proximal Algorithm (PPXA) for minimizing the sum of proximable functions
- ConsensusADMM for solving the global consensus problem with ADMM
- add tests in test_solver.py
  - test_ConsensusADMM_with_ADMM
  - test_ConsensusADMM_with_ADMM_for_Lasso
  - test_ConsensusADMM_with_GPG
  - test_PPXA_with_ADMM
  - test_PPXA_with_GPG

mrava87 commented:
Thanks @tttamaki, will take a look soon and provide my review 😄

mrava87 commented on pyproximal/optimization/primal.py:
```diff
@@ -1793,3 +1793,295 @@ def DouglasRachfordSplitting(
+
+        * :math:`\mathbf{p}_{i}^{(k)} = \prox_{\frac{\tau}{w_i} f_i} (\mathbf{y}_{i}^{(k)})`
+
+      * :math:`\mathbf{p}^{(k)} = \sum_{i=1}^{m} w_i \mathbf{p}_{i}^{(k)}`
```
take away `()` from the superscripts to be consistent with the Notes of other solvers

mrava87 commented on pyproximal/optimization/primal.py:
```diff
@@ -1793,3 +1793,295 @@ def DouglasRachfordSplitting(
+    ----------
+    prox_ops : :obj:`list`
+        A list of proximable functions :math:`f_1, \ldots, f_m`.
+    x0 : :obj:`np.ndarray` or :obj:`list`
```
`x0 : :obj:numpy.ndarray or :obj:list` and also for others below

mrava87 commented on pyproximal/optimization/primal.py:
```diff
@@ -1793,3 +1793,295 @@ def DouglasRachfordSplitting(
+
+    ncp = get_array_module(x0)
+
+    m = len(prox_ops)
```
Add some comments... I guess this all the way to the for loop is `# initialize model`?

mrava87 commented on pyproximal/optimization/primal.py:
```diff
@@ -1793,3 +1793,295 @@ def DouglasRachfordSplitting(
+    prox_ops : :obj:`list`
+        A list of proximable functions :math:`f_1, \ldots, f_m`.
+    x0 : :obj:`np.ndarray` or :obj:`list`
+        Initial vector :math:`\mathbf{x}` of all :math:`f_i` if 1D :obj:`np.ndarray`,
```
Initial vector :math:`\mathbf{x}` for all :math:`f_i` if 1-dimensional array is provided, or initial vectors :math:`\mathbf{x}_{i}` for each :math:`f_i` for :math:`i=1,\ldots,m` if a :obj:`list` of 1-dimensional arrays or a 2-dimensional array of size ** is provided`

** write the size as otherwise it is not obvious where is the dimension of the model vs where is m

tttamaki replied:
I'll fix it, but please modify the part of 2D array shape if nesessary.

mrava87 commented on pyproximal/optimization/primal.py:
```diff
@@ -1793,3 +1793,295 @@ def DouglasRachfordSplitting(
+        Weights :math:`\sum_{i=1}^m w_i = 1, \ 0 < w_i < 1`,
+        Defaults to None, which means :math:`w_1 = \cdots = w_m = \frac{1}{m}.`
+    niter : :obj:`int`, optional
+        The maximum number of iterations.
```
Try to use as much as possible the same description used for the same parameters in other solvers. Here: `Number of iterations of iterative scheme`

mrava87 commented on pyproximal/optimization/primal.py:
```diff
@@ -1793,3 +1793,295 @@ def DouglasRachfordSplitting(
+
+
+def PPXA(  # pylint: disable=invalid-name
+    prox_ops: List[ProxOperator],
```
To be consistent with GPG I suggest to call this `proxfs`

mrava87 commented on pyproximal/optimization/primal.py:
```diff
@@ -1793,3 +1793,295 @@ def DouglasRachfordSplitting(
+        Initial vector :math:`\mathbf{x}` of all :math:`f_i` if 1D :obj:`np.ndarray`,
+        or :math:`\mathbf{x}_{i}` of each :math:`f_i` for :math:`i=1,\ldots,m`
+        if :obj:`list` or 2D :obj:`np.ndarray`.
+    tau : :obj:`float`
```
Is there any condition on tau that must be followed (usually there is an upper bound related to some Lipschitz constant?)

tttamaki replied:
The condition on tau is positive only.
I think that tau is related to gradient, so an upper bound is used in ProximalGradient and GeneralizedProximalGradient. In contrast, methods that use prox only, such as ProximalPoint and DouglasRachfordSplitting (and also PPXA and ConsensusADMM), do not use graident. Hence tau doesn't have an upper bound.

mrava87 commented on pytests/test_solver.py:
```diff
@@ -224,3 +228,220 @@ def test_ADMM_DRS(par):
+        x0=np.zeros(m), tau=tau, niter=150, show=True)
+    xppxa = PPXA(
+        [l2_1, l2_2, l1_1, l1_2],
+        x0=np.zeros(m), tau=tau, niter=15000, show=True, tol=1e-7)
```
That is a lot of iterations.. why do you need so many to match the result of `GeneralizedProximalGradient`?

tttamaki replied:
I fix it because niter=15000 is a tentative value for debuggin, while the iteration stops much earlier by the specified tol.

I notice that I have not checked niter values for this and other tests, so I modifid smaller niter for each test with some experiments using 500 sees (0 to 499).

mrava87 commented on pytests/test_solver.py:
```diff
@@ -224,3 +228,220 @@ def test_ADMM_DRS(par):
+    l1_op = L1(sigma=lmd)
+
+    # Step size
+    L = (Rop_stack.H * Rop_stack).eigs(1).real.item()
```
So here you seem to use the same tau, does that mean that similar conditions to those in ADMM apply also to ConsensusADMM?

tttamaki replied:
No, in theory tau has no upper bound as I replied in other comment. So I change the test code with random tau. However in practice I found that too large tau doesn't pass some tests with a particular seed.

mrava87 reviewed (changes requested):
@tttamaki great work! I just left some minor comments but in general I am happy with the new additions 😄

tttamaki committed:
PR of PPXA and ConsensusADMM refined

primal.py
- prox_ops --> proxfs
- np --> numpy
- fix description of niter
- remove () in math superscripts
- add some comments

test_solver.py
- change niter to fit each test with seeds 0 to 499
- change tau to random value
- change the order of argments to assert_array_almost_equal (actual <--> desired)

tttamaki commented:
@mrava87 Thanks for a quick merge! Now I can use these solvers as official ones :-)
